### PR TITLE
Search pagination 3

### PR DESCRIPTION
**Pull request by AVakiliT**
Since the view classes don't make sense anymore, given how mark wants to implement pagination, I refactored them into a single class with a get method.
This also means we no longer have to read SERPs from the database. 

Anserini now throws away top documents when it paginates. Like Mark said results are cached so there isn't a performance hit since I'm guessing Lucene does some caching in the background.

I also added an elastic search engine interface. I indexed athome4 with anserinis elastic integration and use the elastic api to search the docs. Still don't understand why we aren't using this method in production. Maybe the data collection is too large? Elastic also implements pagination and snippets out of the box so I added them as well. 

To use elastic, I added another docker-compose file called elastic.yml to root that uses elastic for search instead. To use elastic you need to change the interface file path and ip in the settings file as well. Once we settle on how we are going to integrate elastic we can probably figure out a more elegant solution. 

I also don't think it's necessary to store SERP contents and snippets fields in the database, so I removed them when logging SERPs.

### CODE REVIEW

**Comment by KamyarGhajar** [comment]
Nice. Was there a specific reason that you are using `from` for pagination and not the `scroll`? @AVakiliT

**Comment by AVakiliT** [comment]
> Nice. Was there a specific reason that you are using `from` for pagination and not the `scroll`? @AVakiliT

I agree we should eventually switch to scroll if we're expecting users to be doing deep paginations. I just wanted to get it working without having to worry about elastic specific stuff.

**Comment by KamyarGhajar** [comment]
Okay, another thing is that why we are having a docker file for elastic search? there is an ES docker image already there on ELK registry that we should have as a stand-alone micro-service. That does not concern the Gathera project I guess. I will set it up and index the desired docs so everyone can access it online on the deployment server that its address should be set in the config. For the docker_compose file, I have the same comment.

**Comment by AVakiliT** [comment]
Its just for testing with the athome4 collection. It doesn't matter where elastic is since all we need is its ip/hostname.

**Comment by ammsa** [comment]
Logging the SERP is helpful later when we want to do some user behaviour analysis. It makes it easier to know what a user has looked it. Let's keep it for now atleast.

It seems like each search page request creates a new queryID? It will be nice to have it still be associated to the same queryID and append to the SERP field accordingly with the new page results.

Can you upload the ES index so that I can try it locally?

**Comment by AVakiliT** [comment]
> Logging the SERP is helpful later when we want to do some user behaviour analysis. It makes it easier to know what a user has looked it. Let's keep it for now atleast.
Oh okay, I just assumed we could query the index if we needed those. I'll change it back.
> It seems like each search page request creates a new queryID? It will be nice to have it still be associated to the same queryID and append to the SERP field accordingly with the new page results.
Do you think we should append to the existing SERP field in the DB or have a one-to-many relationship between query and SERPs? I don't think merging SERPs in python and rewriting to the DB will be to much of  a performance hit but I'll look into it.
> Can you upload the ES index so that I can try it locally?
Sure I'll upload to dropbox or something and attach a link.

**Comment by AVakiliT** [comment]
Put the esdata folder in data and start using elastic.yml. Remember to change search ip in config to 9200 and also change the search interface class.
[https://www.dropbox.com/s/j9ym9f5zt5c0ex8/esdata.7z?dl=0](url)

**Comment by AVakiliT** [comment]
Queries are now unique within a users session.

So I'm not sure if the user revisits a search page for a query we should log a new SERP in the DB. Right now it doesn't. i.e. a SERP is unique for num_display, page_number and query_id.

There's also a windows specific bug in the docker-compose file which is why I think linh nhi was having trouble getting gathera up and running, relative volume paths need to start with a "./".

Since I added two fields to the search_result model, there might be some DB errors so you might have to delete the db or at least the search_results table.

I also wanted to ask how do we show errors to the user? For example if the search engine is offline and we get a 404 or 500 from the search service how should the error be shown. Is there a specific web page?

**Comment by KamyarGhajar** [comment]
@ammsa @AVakiliT Guys the ES index is usable on the cloud server on cluster ports 9201, 9202, 9203. The ports are not accessible from outside of the server due to firewall policies, but you can deploy gathera on the cloud server where the ES cluster is accessible. Give it a try for warc docs with index name: `warc`.

**Comment by AVakiliT** [comment]
> @ammsa @AVakiliT Guys the ES index is usable on the cloud server on cluster ports 9201, 9202, 9203. The ports are not accessible from outside of the server due to firewall policies, but you can deploy gathera on the cloud server where the ES cluster is accessible. Give it a try for warc docs with index name: `warc`.

Nice, after this is merged we can make a new branch where we switch over to elastic.

**Comment by KamyarGhajar** [comment]
> > @ammsa @AVakiliT Guys the ES index is usable on the cloud server on cluster ports 9201, 9202, 9203. The ports are not accessible from outside of the server due to firewall policies, but you can deploy gathera on the cloud server where the ES cluster is accessible. Give it a try for warc docs with index name: `warc`.
> 
> Nice, after this is merged we can make a new branch where we switch over to elastic.

The index name with 2 replicas and 32 shards is changed to `commoncrawl`

**Comment by KamyarGhajar** [comment]
@AVakiliT @ammsa  It is better to use [Elasticsearch python client](https://elasticsearch-py.readthedocs.io/en/master/) instead of the raw HTTP API calls. It handles the replica round-robin calling, too.

**Comment by ammsa** [comment]
I tried the elastic container locally. Seems to be working, but after some time, it crashed for some reason. Possible because I'm running out of memory in my machine.

> I think I understand what you meant now, I shouldn't create duplicate SERPs or Queries. Should I add page_number and num_display as fields to SERP DB objects?

I don't have any particular way. Whatever you feel is suitable is fine with me! 

> I also wanted to ask how do we show errors to the user? For example if the search engine is offline and we get a 404 or 500 from the search service how should the error be shown. Is there a specific web page?

I think you can raise `Http404`. Something like 
```
from django.http import Http404
...

try:
    ....
except Exception:
    raise Http404("Search error..")
```

We should probably switch to the Elasticsearch python client Kamyar has mentioned.

File: web/Web/web/search/templates/search/search_list.html [hunk 1]
```diff
@@ -82,31 +85,36 @@ <h6 class="title" style="margin-bottom: 0;">{{ hit.title|escape }}</h6>
...
-    <li class="page-item {{ pagination.is_last_page|yesno:'disabled,' }}">
-        <a class="page-link" href="{% url 'search:query_result' query_id=queryID%}?page_number={{pagination.last_page}}">Last</a>
+        <a class="page-link" href="{% url 'search:main'%}?query={{query.split|join:'+'}}&page_number={{pagination.page_number|add:+1}}">Next</a>
     </li>
 
   </ul>
 </nav>
+
+<style>
+  .document em {
```
Comment (ammsa) [inline web/Web/web/search/templates/search/search_list.html:113]: Let's move this to `static/css/project.css`. https://github.com/UWaterlooIR/gathera/blob/06d8b3e05907a6af6f47439288271f7b96531d9a/web/Web/web/static/css/project.css#L710

**Comment by AVakiliT** [comment]
If Elastic crashed it might be because of memory issues. I tried to limit it to a single node to reduce memory usage. It shouldn't be a problem in production. 

After this is merged do I'll change the defaults settings and try and get this working on the cloud server. Do I have enough privileges to stop the and restart the current instance of gathera?

I'll probably create another branch to fix error handling since we should get this merged before Wednesday.

File: web/Web/web/search/views.py [hunk 2]
```diff
@@ -28,115 +26,86 @@
...
-        query_instance = Query.objects.create(
-            username=request.user,
-            query=search_input,
-            session=request.user.current_session,
-        )
-
-        # Call search API to get search results
-        SERP = SearchEngine.search(search_input)
+        query = request.GET.get('query', '')
+        if query != '':
```
Comment (ammsa) [inline web/Web/web/search/views.py:67]: Minor edit. Change to 
```
query = request.GET.get('query', None)
if query:
```

**Comment by KamyarGhajar** [comment]
> I tried the elastic container locally. Seems to be working, but after some time, it crashed for some reason. Possible because I'm running out of memory in my machine.
> 
> > I think I understand what you meant now, I shouldn't create duplicate SERPs or Queries. Should I add page_number and num_display as fields to SERP DB objects?
> 
> I don't have any particular way. Whatever you feel is suitable is fine with me! But let's keep queries unique within a users session.
> 
> > I also wanted to ask how do we show errors to the user? For example if the search engine is offline and we get a 404 or 500 from the search service how should the error be shown. Is there a specific web page?
> 
> I think you can raise `Http404`. Something like
> 
> ```
> from django.http import Http404
> ...
> 
> try:
>     ....
> except Exception:
>     raise Http404("Search error..")
> ```
> 
> We should probably switch to the Elasticsearch python client Kamyar has mentioned.

If you were checking the ES cluster, I was setting up SSL with passwords on it, so it won't respond without correct privileges now.

File: web/Web/web/interfaces/SearchEngine/elastic.py [hunk 3]
```diff
@@ -0,0 +1,81 @@
+import json
+
+import requests
+from config.settings.base import SEARCH_SERVER_IP
+from config.settings.base import SEARCH_SERVER_PORT
+from config.settings.base import INDEX_NAME
+from config.settings.base import DEFAULT_NUM_DISPLAY
+from web.interfaces.SearchEngine.base import SearchInterface
+from collections import OrderedDict
```
Comment (ammsa) [inline web/Web/web/interfaces/SearchEngine/elastic.py:9]: What's `OrderedDict` for?

**Comment by KamyarGhajar** [comment]
@AVakiliT @ammsa With the new privileges on the ES cluster on the server you should use username/password to connect to ES for search. Rather than setting the credentials hard-coded on the git files, you should set config files with ES credentials on the deployment environment. If you need credentials I have created one for the gather server.

File: web/Web/web/search/models.py [hunk 4]
```diff
@@ -1,12 +1,14 @@
...
 from config.settings.base import AUTH_USER_MODEL as User
+from config.settings.base import DEFAULT_NUM_DISPLAY
 from django.contrib.postgres.fields import JSONField
 from django.db import models
-
+from django.core.validators import MaxValueValidator, MinValueValidator
 from web.core.models import Session
 
 
+
```
Comment (ammsa) [inline web/Web/web/search/models.py:11]: Two blank lines after import statements just to follow PEP8 guidelines.

**Comment by AVakiliT** [comment]
@ammsa Made the changes, Thanks for the review 👍.

**Comment by ammsa** [comment]
Looks good! Merging.

**Comment by ammsa** [comment]
In a different branch, we can look into using a Elasticsearch python client.

**Pull request merged**